lib/shelley/src/Cardano/Wallet/Shelley.hs

```diff
@@ -0,0 +1,464 @@
+    ( IcarusKey )
+import Cardano.Wallet.Primitive.AddressDiscovery
+    ( IsOurs )
+import Cardano.Wallet.Primitive.AddressDiscovery.Random
+    ( RndState )
+import Cardano.Wallet.Primitive.AddressDiscovery.Sequential
+    ( SeqState )
+import Cardano.Wallet.Primitive.Types
+    ( Address
+    , Block
+    , BlockchainParameters (..)
+    , ChimericAccount
+    , GenesisBlockParameters (..)
+    , SyncTolerance
+    , WalletId
+    )
+import Cardano.Wallet.Registry
+    ( WorkerLog (..) )
+import Cardano.Wallet.Shelley.Api.Server
+    ( server )
+import Cardano.Wallet.Shelley.Compatibility
+    ( Shelley, ShelleyBlock, fromNetworkMagic, fromShelleyBlock )
+import Cardano.Wallet.Shelley.Network
+    ( NetworkLayerLog, withNetworkLayer )
+import Cardano.Wallet.Shelley.Transaction
+    ( newTransactionLayer )
+import Cardano.Wallet.Shelley.Transaction.Size
+    ( MaxSizeOf )
+import Cardano.Wallet.Transaction
+    ( TransactionLayer )
+import Control.Applicative
+    ( Const (..) )
+import Control.Tracer
+    ( Tracer (..), nullTracer, traceWith )
+import Data.Function
+    ( (&) )
+import Data.Proxy
+    ( Proxy (..) )
+import Data.Text
+    ( Text )
+import Data.Text.Class
+    ( ToText (..) )
+import GHC.Generics
+    ( Generic )
+import Network.Ntp
+    ( NtpClient (..), NtpTrace (..), withWalletNtpClient )
+import Network.Socket
+    ( SockAddr, Socket, getSocketName )
+import Network.Wai.Handler.Warp
+    ( setBeforeMainLoop )
+import Network.Wai.Middleware.Logging
+    ( ApiLog )
+import Ouroboros.Consensus.Node.Run
+    ( RunNode )
+import Ouroboros.Consensus.Shelley.Ledger
+    ( Crypto )
+import Ouroboros.Consensus.Shelley.Protocol.Crypto
+    ( TPraosCrypto )
+import Ouroboros.Network.CodecCBORTerm
+    ( CodecCBORTerm )
+import Ouroboros.Network.NodeToClient
+    ( NodeToClientVersionData (..) )
+import System.Exit
+    ( ExitCode (..) )
+import System.IOManager
+    ( withIOManager )
+
+import qualified Cardano.Wallet.Api.Server as Server
+import qualified Cardano.Wallet.DB.Sqlite as Sqlite
+import qualified Data.Text as T
+import qualified Network.Wai.Handler.Warp as Warp
+
+-- | Encapsulate a network discriminant and the necessary constraints it should
+-- satisfy.
+data SomeNetworkDiscriminant where
+    SomeNetworkDiscriminant
+        :: forall (n :: NetworkDiscriminant).
+            ( NetworkDiscriminantVal n
+            , PaymentAddress n IcarusKey
+            , PaymentAddress n ByronKey
+            , DecodeAddress n
+            , EncodeAddress n
+            , MaxSizeOf Address n IcarusKey
+            , MaxSizeOf Address n ByronKey
+            )
+        => Proxy n
+        -> SomeNetworkDiscriminant
+
+deriving instance Show SomeNetworkDiscriminant
+
+-- | The @cardano-wallet-shelley@ main function. It takes the configuration
+-- which was passed from the CLI and environment and starts all components of
+-- the wallet.
+serveWallet
+    :: forall t c.
+        ( t ~ IO (Shelley c)
```

Comment:
Is there any value for us to carry this `c` parameter around? Certainly, we can already make a choice for the underlying Crypto scheme (for instance: `TPraosStandardCrypto`) as it seems to be done in:

https://github.com/input-output-hk/cardano-node/blob/master/cardano-api/src/Cardano/Api/Types.hs#L127-L146

Reply:
Was wondering the same. Went with it to see how it would work out. Maybe we would want to use some mock crypto for… performance at some point?

But certainly we can remove it.

Reply:
When the crypto becomes our bottleneck in performances here, we would have come a long way :grimacing: